In [5]:
import pandas as pd 
import matplotlib.pyplot as plt 
import numpy as np 
import seaborn as sns

In [6]:
df = pd.read_csv('Cleaned_data.csv')

In [7]:
df.head()

Unnamed: 0                  Job Title Salary Estimate  \
0           0             Data Scientist          53-91    
1           1  Healthcare Data Scientist         63-112    
2           2             Data Scientist          80-90    
3           3             Data Scientist          56-97    
4           4             Data Scientist         86-143    

                                     Job Description  Rating  \
0  Data Scientist\nLocation: Albuquerque, NM\nEdu...     3.8   
1  What You Will Do:\n\nI. General Summary\n\nThe...     3.4   
2  KnowBe4, Inc. is a high growth information sec...     4.8   
3  *Organization and Job ID**\nJob ID: 310709\n\n...     3.8   
4  Data Scientist\nAffinity Solutions / Marketing...     2.9   

                            Company Name         Location    Headquarters  \
0                      Tecolote Research  Albuquerque, NM      Goleta, CA   
1  University of Maryland Medical System    Linthicum, MD   Baltimore, MD   
2                                KnowBe4   Clearwater, FL  Clearwater, FL   
3                                   PNNL     Richland, WA    Richland, WA   
4                     Affinity Solutions     New York, NY    New York, NY   

                     Size  Founded  ... age python  R spark aws  excel  \
0   501 to 1000 employees     1973  ...  49      1  0     0   0      1   
1        10000+ employees     1984  ...  38      1  0     0   0      0   
2   501 to 1000 employees     2010  ...  12      1  0     1   0      1   
3  1001 to 5000 employees     1965  ...  57      1  0     0   0      0   
4     51 to 200 employees     1998  ...  24      1  0     0   0      1   

         Job_simp  seniority  num_comp  desc_len  
0  data scientist         na         1      2536  
1  data scientist         na         1      4783  
2  data scientist         na         1      3461  
3  data scientist         na         3      3883  
4  data scientist         na         3      2728  

[5 rows x 32 columns]

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 742 entries, 0 to 741
Data columns (total 32 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         742 non-null    int64  
 1   Job Title          742 non-null    object 
 2   Salary Estimate    742 non-null    object 
 3   Job Description    742 non-null    object 
 4   Rating             742 non-null    float64
 5   Company Name       742 non-null    object 
 6   Location           742 non-null    object 
 7   Headquarters       742 non-null    object 
 8   Size               742 non-null    object 
 9   Founded            742 non-null    int64  
 10  Type of ownership  742 non-null    object 
 11  Industry           742 non-null    object 
 12  Sector             742 non-null    object 
 13  Revenue            742 non-null    object 
 14  Competitors        742 non-null    object 
 15  hourly             742 non-null    int64  
 16  employer_provided  742 non

In [80]:
df_model = df[['avg_salary','Size','Type of ownership','Industry','num_comp',
        'job_state','age','python','excel','Job_simp',]]

**columns to be used for model building**


1.   avg_salary(dependent feature)
2.   Type of ownership
              Company - Private                 410
              Company - Public                  193
              Nonprofit Organization             55
              Subsidiary or Business Segment     34
              Government                         15
              Hospital                           15
              College / University               13
              Other Organization                  3
              School / School District            2
              Unknown                             1 

3.   Size (  Number of employes)

          001 to 5000 employees     150
          501 to 1000 employees     134
          10000+ employees          130
          201 to 500 employees      117
          51 to 200 employees        94
          5001 to 10000 employees    76
          1 to 50 employees          31
          Unknown                     9 
4. Industry
5. Sector
6. num_comp( Number of competators)

          1    472
          3    228
          2     41
          4      1 
7. job_state (38 states)
8. same_location( check weather job state is  same that of  headquaters)
9. age (Number of years since the company is present in market)
10. python (if job role consists of python in description)
11. excel (if job role consists of excel in description)
12. Job_simp ( job positions )
          data scientist    279
          na                184
          data engineer     119
          analyst           102
          manager            22
          mle                22
          director           14

In [10]:
df_model.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 742 entries, 0 to 741
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   avg_salary         742 non-null    float64
 1   Size               742 non-null    object 
 2   Type of ownership  742 non-null    object 
 3   Industry           742 non-null    object 
 4   Sector             742 non-null    object 
 5   num_comp           742 non-null    int64  
 6   job_state          742 non-null    object 
 7   same_location      742 non-null    int64  
 8   age                742 non-null    int64  
 9   python             742 non-null    int64  
 10  excel              742 non-null    int64  
 11  Job_simp           742 non-null    object 
dtypes: float64(1), int64(5), object(6)
memory usage: 69.7+ KB


In [81]:
# get dummy data 
df_dum = pd.get_dummies(df_model)


In [82]:
X = df_dum.drop('avg_salary', axis= 'columns')
y = df_dum['avg_salary']

In [83]:
# train test split 
from sklearn.model_selection import train_test_split

In [84]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [85]:
#Lasso model

from sklearn.linear_model import Lasso
from sklearn.model_selection import cross_val_score

In [86]:
lm_l = Lasso(alpha=.13)
lm_l.fit(X_train,y_train)
np.mean(cross_val_score(lm_l,X_train,y_train, scoring = 'neg_mean_absolute_error', cv= 3))


-22.867078903517182

In [104]:
#RandomForestRegressor model

from sklearn.ensemble import RandomForestRegressor
rf = RandomForestRegressor()

np.mean(cross_val_score(rf,X_train,y_train,scoring = 'neg_mean_absolute_error', cv= 3))

-17.116504761646798

In [88]:
from sklearn.model_selection import RandomizedSearchCV

In [184]:
 #Randomized Search CV

# Number of trees in random forest
n_estimators = [int(x) for x in np.linspace(start = 100, stop = 1200, num = 12)]
# Number of features to consider at every split
max_features = ['auto', 'sqrt']
# Maximum number of levels in tree
max_depth = [int(x) for x in np.linspace(5, 30, num = 6)]
# max_depth.append(None)
# Minimum number of samples required to split a node
min_samples_split = [2, 5, 10, 15, 100]
# Minimum number of samples required at each leaf node
min_samples_leaf = [1, 2, 5, 10,15,20]

In [185]:
# Create the random grid
random_grid = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf}

print(random_grid)

{'n_estimators': [100, 200, 300, 400, 500, 600, 700, 800, 900, 1000, 1100, 1200], 'max_features': ['auto', 'sqrt'], 'max_depth': [5, 10, 15, 20, 25, 30], 'min_samples_split': [2, 5, 10, 15, 100], 'min_samples_leaf': [1, 2, 5, 10, 15, 20]}


In [190]:
# Random search of parameters, using 3 fold cross validation, 
# search across 100 different combinations
rf_random = RandomizedSearchCV(estimator = rf, param_distributions = random_grid, n_iter = 10, cv = 5, verbose=2, random_state=42, n_jobs = 1)


In [205]:
rf_random.fit(X_train,y_train)

Fitting 5 folds for each of 10 candidates, totalling 50 fits
[CV] END max_depth=10, max_features=auto, min_samples_leaf=5, min_samples_split=5, n_estimators=900; total time=   2.6s
[CV] END max_depth=10, max_features=auto, min_samples_leaf=5, min_samples_split=5, n_estimators=900; total time=   2.3s
[CV] END max_depth=10, max_features=auto, min_samples_leaf=5, min_samples_split=5, n_estimators=900; total time=   2.3s
[CV] END max_depth=10, max_features=auto, min_samples_leaf=5, min_samples_split=5, n_estimators=900; total time=   2.2s
[CV] END max_depth=10, max_features=auto, min_samples_leaf=5, min_samples_split=5, n_estimators=900; total time=   2.3s
[CV] END max_depth=30, max_features=auto, min_samples_leaf=5, min_samples_split=100, n_estimators=500; total time=   1.0s
[CV] END max_depth=30, max_features=auto, min_samples_leaf=5, min_samples_split=100, n_estimators=500; total time=   1.0s
[CV] END max_depth=30, max_features=auto, min_samples_leaf=5, min_samples_split=100, n_estimato

RandomizedSearchCV(cv=5, estimator=RandomForestRegressor(), n_jobs=1,
                   param_distributions={'max_depth': [5, 10, 15, 20, 25, 30],
                                        'max_features': ['auto', 'sqrt'],
                                        'min_samples_leaf': [1, 2, 5, 10, 15,
                                                             20],
                                        'min_samples_split': [2, 5, 10, 15,
                                                              100],
                                        'n_estimators': [100, 200, 300, 400,
                                                         500, 600, 700, 800,
                                                         900, 1000, 1100,
                                                         1200]},
                   random_state=42, verbose=2)

In [206]:
rf_random.best_params_


{'max_depth': 25,
 'max_features': 'auto',
 'min_samples_leaf': 1,
 'min_samples_split': 15,
 'n_estimators': 400}

In [207]:
rf_random.score(X_test,y_test)

0.6402610135705411

In [208]:
predictions = rf_random.predict(X_test)

In [209]:
from sklearn import metrics

In [210]:
print('MAE:', metrics.mean_absolute_error(y_test, predictions))


MAE: 17.85213888482423


This is the maximum model accuracy I am able to achieve with difference of 17k 

In [227]:
import pickle
file = open('Salary_predictions.pkl', 'wb')
pickle.dump(rf_random, file)